File: app/src/main/java/me/zhanghai/android/files/provider/linux/LinuxFileAttributeView.java
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.linux;

import android.os.Parcel;

import java.util.Set;

import androidx.annotation.NonNull;
import java8.nio.file.Path;
import me.zhanghai.android.files.provider.root.RootPosixFileAttributeView;
import me.zhanghai.android.files.provider.root.RootablePosixFileAttributeView;

public class LinuxFileAttributeView extends RootablePosixFileAttributeView {

    static final Set<String> SUPPORTED_NAMES = LocalLinuxFileAttributeView.SUPPORTED_NAMES;

    @NonNull
    private final LinuxPath mPath;
    private final boolean mNoFollowLinks;

    LinuxFileAttributeView(@NonNull LinuxPath path, boolean noFollowLinks) {
        super(path, new LocalLinuxFileAttributeView(path.toByteString(), noFollowLinks),
                RootPosixFileAttributeView::new);

        mPath = path;
        mNoFollowLinks = noFollowLinks;
    }


    public static final Creator<LinuxFileAttributeView> CREATOR =
            new Creator<LinuxFileAttributeView>() {
                @Override
                public LinuxFileAttributeView createFromParcel(Parcel source) {
                    return new LinuxFileAttributeView(source);
                }
                @Override
                public LinuxFileAttributeView[] newArray(int size) {
                    return new LinuxFileAttributeView[size];
                }
            };

    protected LinuxFileAttributeView(Parcel in) {
        this(in.readParcelable(Path.class.getClassLoader()), in.readByte() != 0);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeParcelable(mPath, flags);
        dest.writeByte(mNoFollowLinks ? (byte) 1 : (byte) 0);
    }
}

File: app/src/main/java/me/zhanghai/android/files/provider/linux/LinuxFileAttributeView.kt
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.provider.linux

import android.os.Parcel
import android.os.Parcelable
import me.zhanghai.android.files.compat.readBooleanCompat
import me.zhanghai.android.files.compat.writeBooleanCompat
import me.zhanghai.android.files.provider.root.RootPosixFileAttributeView
import me.zhanghai.android.files.provider.root.RootablePosixFileAttributeView

internal class LinuxFileAttributeView constructor(
    private val path: LinuxPath,
    private val noFollowLinks: Boolean
) : RootablePosixFileAttributeView(
    path, LocalLinuxFileAttributeView(path.toByteString(), noFollowLinks),
    { RootPosixFileAttributeView(it) }
) {
    private constructor(source: Parcel, loader: ClassLoader?) : this(
        source.readParcelable(loader)!!, source.readBooleanCompat()
    )

    override fun describeContents(): Int = 0

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeParcelable(path, flags)
        dest.writeBooleanCompat(noFollowLinks)
    }

    companion object {
        val SUPPORTED_NAMES = LocalLinuxFileAttributeView.SUPPORTED_NAMES

        @JvmField
        val CREATOR = object : Parcelable.ClassLoaderCreator<LinuxFileAttributeView> {
            override fun createFromParcel(source: Parcel): LinuxFileAttributeView =
                createFromParcel(source, null)

            override fun createFromParcel(
                source: Parcel,
                loader: ClassLoader?
            ): LinuxFileAttributeView = LinuxFileAttributeView(source, loader)

            override fun newArray(size: Int): Array<LinuxFileAttributeView?> = arrayOfNulls(size)
        }
    }
}
